protocol/transfer-precompile.md

```diff
@@ -0,0 +1,83 @@
+# Purpose
+
+Since its inception, the Celo blockchain provides a feature called "[Token
+Duality](https://specs.celo.org/token_duality.html)", which allows the native
+token to be used both in the usual ETH-like way and as an ERC-20 token.
+
+We would like to bring this feature to more OP-Stack based chains and bring the
+Celo L2 closer to vanilla OP-stack by adding the underlying transfer precompile to
+op-geth.
+
+# Summary
+
+Add a precompile to op-geth that allows transferring native tokens.
+This precompile can only be called by a single privileged contract.
+
+# Problem Statement + Context
+
+Many users and contracts have to deal with both ERC20 tokens and native tokens.
+Since they can't be accessed by the same interface, they either have to be
+handled as two different cases or be explicitly wrapped by an ERC-20 token
+wrapping contract.
+Wrapping tokens is not an ideal solution either, since the wrapped token is
+of limited use (e.g. can't be used to pay gas) and wrapping/unwrapping is
+annoying and incurs additional gas charges.
+
+This can be solved by having a contract that provides an ERC-20 interface
+directly to the native token. To implement such a contract, the blockchain must
+provide a way to transfer native tokens on behalf of the transaction sender.
+
+# Proposed Solution
+
+Add one new blockchain parameter `TransferPrecompileAllowedCaller` and a new precompile.
+The parameter defines the address from which the precompile must be called.
+
+Precompile name: `transfer`  
+Precompile address: 0x0000000000000000000000000000000000000101 (directly after the `P256VERIFY` precompile from Fjord)  
+Parameters (abi-encoded): `address from, address to, uint256 value`  
+Gas costs: 9000  
+Result: `value` native tokens will be transferred from `from` to `to`.  
+No return value  
+
+The precompile can fail for the following reasons:
+* `TransferPrecompileAllowedCaller` is not configured for the blockchain
+* It is not called from the contract at the `TransferPrecompileAllowedCaller` address
```

Comment:
Is the `TransferPrecompileAllowedCaller` address an arbitrary enshrined smart contract? In the context of the OP Stack, this could be a predeploy. Why can't anybody call this contract and instead of `msg.sender` being the first arg, the native EVM can look at the `CALLER` (which is `msg.sender`)?

Reply:
Ok given the [geth code](https://github.com/ethereum/go-ethereum/blob/master/core/vm/contracts.go#L195) and your comment:

> Removing the `from` parameter from the precompile is not an option because 
`msg.sender` can't be accessed from inside the precompile.
There, the caller is set to the contract doing `TRANSFER.call`.

I understand this design decision more. It would be a larger diff to send through the evm context into the precompile

Reply:
> I understand this design decision more. It would be a larger diff to send through the evm context into the precompile

Yes, and it would also reduce the chances of the precompile being supported by third party tooling (e.g. foundry's anvil).